### PR TITLE
perf(db) improve kong startup time by reusing luasocket on kong.init

### PR DESCRIPTION
**Pull request by bungle**
### Summary

This is the seond PR that comes from the discussion of #4171.

Without this patch the Postgres strategy creates 7 non-pooled luasocket connections to database on Kong.init(). This patch reduces that to 3.

Without this patch the Cassandra strategy creates 9 non-pooled luasocket connections to database on Kong.init(). This patch reduces that to 1.

The difference is because Postgres needs to run initialization SQL to setup the connection timezone and database schema on new connections. And for Postgres `connect_migrations` is the same
as `connect` while with Cassandra it is not. It is still a tiny improvement for Postgres as well.

### CODE REVIEW

File: kong/init.lua [hunk 1]
```diff
@@ -344,19 +345,19 @@ function Kong.init()
   sort_plugins_for_execution(config, db, loaded_plugins)
 
   local err
-  plugins_map_semaphore, err = semaphore.new()
+  plugins_map_semaphore, err = semaphore.new(1) -- 1 = treat this as a mutex
   if not plugins_map_semaphore then
     error("failed to create plugins map semaphore: " .. err)
   end
 
-  plugins_map_semaphore:post(1) -- one resource, treat this as a mutex
```
Comment (p0pr0ck5) [inline kong/init.lua]: Curious, why this change? Seems unrelated?

Comment (bungle) [reply kong/init.lua]: @p0pr0ck5 nothing, just a small cleanup as I was touching code close to it. Not a big deal to include or remove the change.

**Comment by jeremyjpj0916** [comment]
Hey folks I gave this a go on 1.0 GA and saw this error:

```
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:259: load_plugin_schemas(): Loading plugin: acl
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:273: load_plugin_schemas(): Loading custom plugin entity: 'acl.acls'
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:259: load_plugin_schemas(): Loading plugin: kong-service-virtualization
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:259: load_plugin_schemas(): Loading plugin: request-transformer
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:259: load_plugin_schemas(): Loading plugin: rate-limiting
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:259: load_plugin_schemas(): Loading plugin: statsd
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:259: load_plugin_schemas(): Loading plugin: oauth2
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:273: load_plugin_schemas(): Loading custom plugin entity: 'oauth2.oauth2_credentials'
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:273: load_plugin_schemas(): Loading custom plugin entity: 'oauth2.oauth2_authorization_codes'
2019/01/13 01:39:13 [debug] 22#0: [lua] plugins.lua:273: load_plugin_schemas(): Loading custom plugin entity: 'oauth2.oauth2_tokens'
2019/01/13 01:39:15 [error] 22#0: init_by_lua error: /usr/local/share/lua/5.1/kong/db/init.lua:125: attempt to concatenate local 'err' (a nil value)
stack traceback:
	/usr/local/share/lua/5.1/kong/db/init.lua:125: in function 'prefix_err'
	/usr/local/share/lua/5.1/kong/db/init.lua:190: in function 'close'
	/usr/local/share/lua/5.1/kong/init.lua:360: in function 'init'
	init_by_lua:3: in main chunk
nginx: [error] init_by_lua error: /usr/local/share/lua/5.1/kong/db/init.lua:125: attempt to concatenate local 'err' (a nil value)
stack traceback:
	/usr/local/share/lua/5.1/kong/db/init.lua:125: in function 'prefix_err'
	/usr/local/share/lua/5.1/kong/db/init.lua:190: in function 'close'
	/usr/local/share/lua/5.1/kong/init.lua:360: in function 'init'
	init_by_lua:3: in main chunk
```

And init.lua line 360 was the assert db close in the newly modified snippet:
```
local err
  plugins_map_semaphore, err = semaphore.new(1) -- 1 = treat this as a mutex
  if not plugins_map_semaphore then
    error("failed to create plugins map semaphore: " .. err)
  end

  local _, err = build_plugins_map(db, "init")
  if err then
    error("error building initial plugins map: ", err)
  end

  assert(runloop.build_router(db, "init"))
  
  assert(db:close())
end
```

Just thought I would bring it to attention since I noticed this was indeed  already merged if it raises any concerns.
